pkg/kubelet/cm/topologymanager/policy_single_numa_node.go

```diff
@@ -52,8 +55,139 @@ func (p *singleNumaNodePolicy) canAdmitPodResult(hint *TopologyHint) lifecycle.P
 	}
 }
 
+// Get the narrowest hint aligned to one socket
+// assumption: allResourcesHints are not empty, all hints have SocketAffinity != nil,
+// all hints have exactly one socket set in SocketAffinity
+func (p *singleNumaNodePolicy) getHintMatch(allResourcesHints [][]TopologyHint) (bool, TopologyHint) {
+	// Sort all resources hints
+	for _, resource := range allResourcesHints {
+		sort.Slice(resource, func(i, j int) bool {
+			if resource[i].NUMANodeAffinity.GetBits()[0] < resource[j].NUMANodeAffinity.GetBits()[0] {
+				return true
+			}
+			return false
+		})
+
+	}
+	// find a match by searching a hint of one resource in the rest
+	var match TopologyHint
+	var foundMatch bool
+
+	if len(allResourcesHints) == 1 {
+		match = allResourcesHints[0][0]
+		match.Preferred = true
+		return true, match
+	}
+	for _, candidate := range allResourcesHints[0] {
+		foundMatch = true
+		for _, hints := range allResourcesHints[1:] {
+			res := sort.Search(len(hints), func(i int) bool {
+				return hints[i].NUMANodeAffinity.GetBits()[0] >= candidate.NUMANodeAffinity.GetBits()[0]
+			})
+			if res >= len(hints) ||
+				!hints[res].NUMANodeAffinity.IsEqual(candidate.NUMANodeAffinity) {
+				// hint not found, move to next hint from allResourcesHints[0]
+				foundMatch = false
+				break
+			}
+		}
+		if foundMatch {
+			match = candidate
+			match.Preferred = true
+			break
+		}
+	}
+	return foundMatch, match
+}
+
+// Return hints that have valid bitmasks with exactly one bit set
+func (p *singleNumaNodePolicy) filterHints(allResourcesHints [][]TopologyHint) [][]TopologyHint {
+	var filteredResourcesHints [][]TopologyHint
+	if len(allResourcesHints) > 0 {
+		for _, oneResourceHints := range allResourcesHints {
+			var filtered []TopologyHint
+			if len(oneResourceHints) > 0 {
+				for _, hint := range oneResourceHints {
+					if hint.NUMANodeAffinity != nil && hint.NUMANodeAffinity.Count() == 1 {
+						filtered = append(filtered, hint)
+					}
+				}
+			}
+			filteredResourcesHints = append(filteredResourcesHints, filtered)
```

Comment:
I think this append will need a check:  `if len(filtered) != 0` otherwise an empty `filtered` could be appended and we will end up breaking out early here: https://github.com/kubernetes/kubernetes/blob/54e5b088d1ded78d27325f61aa4c607033c0cee2/pkg/kubelet/cm/topologymanager/policy_single_numa_node.go#L176